Detail a sprint process for the Enarx project

mbestavros opened the pull request:
You can view the formatted version [here](https://github.com/mbestavros/rfcs/blob/sprints/00003-enarx-sprint-framework/README.md).

Please provide feedback -- discussion on this is important!

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+# Agile Sprint Process for Enarx
+- Authors: [Mark Bestavros](mbestavr@redhat.com) -- email is optional
```
~~email is optional~~

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+# Agile Sprint Process for Enarx
+- Authors: [Mark Bestavros](mbestavr@redhat.com) -- email is optional
+- Status: [PROPOSED](/README.md#proposed)
+- Since: 2020-05-06 (date you submit your PR)
```
~~(date you submit your PR)~~

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Since: 2020-05-06 (date you submit your PR)
+- Status Note: Concept brainstormed, awaiting implementation
+- Supersedes: N/A
+- Start Date: 2020-05-01 (date you started working on this idea)
```
~~(date you started working on this idea)~~

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+### Backlog
+
+As our project scales, it’s easy to imagine a single unified backlog becoming incredibly cumbersome and hard to manage. Instead, we should adapt our current approach to area-specific boards, which effectively divides our backlog into categories (SGX, SEV,  infrastructure, etc.).
```
My first reaction as I read this paragraph was to agree. However, I think there might be some value to consolidating our boards into a single board. The primary reason is that when we're sprinting, I'd argue that the single most important unit of work is our active sprint (comprised of issues that should arguably sum to a reasonable amount of work within a workweek per engineer). By having just one board, we can capture the contents of the sprint with one column on one board, rather than having this fragmented across component-level boards. I think we can also accomplish this with GitHub milestones.

In any case, I don't think smaller boards buy us anything. Even if the rate at which we accumulate work vastly outpaces the work completed per sprint, I think we'd benefit more from disciplined grooming of the backlog and decomposition of tasks than fragmenting the backlog into component-sized boards. We will need to advocate ruthlessly for understanding the priority of individual issues and how they fit in with our roadmap moving forward. Regardless of the number of boards, the size of the backlog is the size of the backlog, regardless of if it's divided by 1 board or N-boards.

lkatalin replied:
I just checked, and we are able to filter cards by labels while on a board, so this could help us stay organized if we do decide to have just one backlog instead of multiple backlogs.

axelsimon replied:
I agree with @connorkuehl here. One issue would sub-boards also arises when you need to work on something that affects multiple components at the same time.

That's not to say they can't be useful, as filtering out stuff to just see what you need to see is clearly useful at times.

Sub-boards as just a filtered "main board" seem like a good way to look at this to me.

MikeCamel replied:
A "Sprint board" is going to be important - I see this noted below.

npmccallum replied:
I am noting consensus on a single sprint board.

mbestavros replied:
Resolving to simplify older comments.

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Done: PRs that have been merged, and their associated issues, go here (automatable with stock Github).
+
+A few things to consider:
+- Advantage: having project board be separate from backlog means we can promote issues to be part of the current sprint whenever we want to, and they can skip the backlog entirely if that’s desired.
```
I think we should be cautious with a mechanism like this. If we abuse it then it undermines the efforts that go into sprint planning.

npmccallum replied:
@connorkuehl Agreed. However, I asked for this explicitly because I want a way to capture drive-by contributions. I don't want a new contributor to have to wait 2 weeks for a PR just because they missed the planning meeting by 5 minutes.

lkatalin replied:
I also have it come up pretty often that something I'm working on (or something that comes up within @npmccallum 's work) spawns a new issue that gets promoted to my immediate workload... so not being able to do that seems strange? Cumbersome? I can think of sprint-compliant ways to do this sort of thing, but we should probably talk about what we want that to look like.

axelsimon replied:
> Agreed. However, I asked for this explicitly because I want a way to capture drive-by contributions. I don't want a new contributor to have to wait 2 weeks for a PR just because they missed the planning meeting by 5 minutes.

This makes sense, but i think drive-by contributors should not be considered to be "part of the team", and as such exempt from following sprints.

@lkatalin According to agile doctrine, this would suggest that the original issue was either not groomed properly, or not broken down into the right subtasks during sprint planning. The spawned issue you run into would likely have emerged as the issue-owner tries to explain the issue to the rest of the team and everyone reaches a common understanding (backlog grooming) or as the issue is broken down into immediately actionable tasks (sprint planning).

This is all a bit theoretical of course, because there's always stuff that just wasn't foreseeable. But then again, the concept of spikes, or stories to explore something that isn't well understood enough to be done as a normal user story exists for this case. :smiley:

lkatalin replied:
@axelsimon Yeah, I think this partially speaks to my (mild) skepticism that some of our workflow can entirely fit into the Platonic ideal "sprint" because it's all new ground and we often don't know what the sub-tasks or blockers are until we try a few different things. But it's also possible that we can be more exact about sub-tasks than we currently are. Only time will tell. I am okay with trying to conform to the sprint and seeing how it goes, and adapting it as required.

npmccallum replied:
As I noted on the call today, I think the right way to move forward here is to treat issues and PRs differently. Sprint planning decides which issues are placed on the sprint board. Issues may or may not have a corresponding PR. All PRs go on the sprint board, regardless of whether they have a corresponding issue or not.

This has the following properties.

1. We prioritize the work of the core team.
2. We can have non-code issues as part of planning. For example, infrastructure issues. These don't need PRs.
3. All PRs have the same process visibility. Whether your work is "core team" work or "outside contributor work" you get the same visibility.
4. Joining the "core team" means you take part in sprint planning. This is open to anyone.

npmccallum replied:
I forgot to mention three things regarding issues.

1. We probably need some sort of issue triage that is more frequent than the Sprint duration. At the very least we want a prompt human touch on each issue that is filed so that contributors feel welcomed. We also need to identify urgent issues...

2. We need to have a notion of "interrupt" issues which bypass sprint planning and go directly to the board even in the middle of a sprint. These should be rare. They should either (1) block existing sprint work or (2) have otherwise catastrophic impact (think: website is down).

3. What is the process for mid-sprint decomposing? What if we have planned an issue for a sprint and while doing it we notice it should be decomposed further? How do we handle this case?

lkatalin replied:
I second that mid-sprint decomposing or some type of check-in re: the chosen work would be beneficial.

MikeCamel replied:
> 3. What is the process for mid-sprint decomposing? What if we have planned an issue for a sprint and while doing it we notice it should be decomposed further? How do we handle this case?

Someone needs to make a call (or a specified group of people, or a process for voting, etc.).  This is another reason that we need to define roles and responsibilities.

mbestavros replied:
@npmccallum My thinking around point 1) of yours above was that we'd do new issue triage _every standup_. This would mean new issues get a human response within 24 working hours. Were you thinking of something in addition to that?

npmccallum replied:
No.

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+For each board:
+- Briefly run through the Backlog column to decide if there are any issues we want to promote to "Tracking" status.
+- Run through all items in "Tracking" and decide which ones are highest priority, and who will work on them. They will become the issue’s assignee.
```
I think it's important that we come up with some agreeable "unit" of work and estimate how many units of work comprise a "full" sprint for an engineer.

Furthermore, when grooming or picking tasks we should have some consensus based on the estimated units of work that issue will take. Just to be clear, this is not a do-or-die deadline (and it's most certainly not to be used _against_ anyone during retro), but this is a tool for helping us not overload someone with work. Estimating scope of most non-trivial issues is exceptionally hard, especially since we have such a component-based group, but I think it's still worthwhile. These metrics may be useful during the retrospective. If the unit is agreed on as a team, that ties in to the fact that we sprint as a team and we ultimately move code as a team. If issues are left unfinished due to scope or difficulty, then at least we can share that learning experience as a team.

lkatalin replied:
I agree with @connorkuehl , I think what a "unit" of work is and what each task is worth should be considered carefully. I will admit up front that I have no idea how long a task might take more than half the time, because I mostly do things I've never done before. Sometimes a task I think will take a day takes three weeks, and sometimes (rarely) a task I think will take two weeks takes an afternoon. Others can probably help out here in estimating, but we also have to keep in mind that something that might take a senior engineer two days could take a more junior engineer two weeks or an intern a month. IMO this is the biggest challenge with moving to a "sprint" structure since so much of our project is uncharted territory and new problems come up en route to even straightforward-seeming issues.

MikeCamel replied:
There are also particular roles to be considered here: it's not just a free-for-all discussion.  It needs to be led and coordinated.

axelsimon replied:
The unit of work is supposed to stem from the "complexity points", AFAIK. See comment above regarding backlog grooming vs. sprint planning on this.

npmccallum replied:
I think work estimation is indeed important. However, I think it is difficult and I'd like for us to ease our way into this process. What do people think about leaving the unit undefined for now so that we can test our workflow and automation? Then we can add in "work units" and more careful planning as a follow-up.

I'm thinking perhaps we outline the workflow first and then implement it. This puts the RFC in the `demonstrated` state. Then we add "work units" before the transition to `adopted`. Thoughts?

lkatalin replied:
I like the idea of leaving the "work unit" undefined at first and getting the rest of the process working, since I think estimating work will be a challenge in itself.

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+Finally, if there's time, we'll dedicate some time to a quick review of our Current Sprint board. This will serve as an opportunity for developers to elaborate on any blockers and ask for reviews or re-reviews as needed. We should strive to do this every day if possible, but if prior discussion makes that infeasible, we don't _have_ to do it (there's always tomorrow). Project discussion is more i ...
+
+### Sprint Retrospective
+Our final meeting is the sprint retrospective. This should be a quick half-hour meeting that takes the place of Monday’s standup where we discuss:
```
I think we should err on the side of caution and block off 60 minutes for the retrospective.

lkatalin replied:
A 60 minute retrospective could make sense. If we do this and the planning meeting is the same day, I would strongly vote to have them back-to-back so that if the retrospective only takes 25 minutes we can immediately transition into planning, instead of having an awkward gap between the two and having two hours blocked off that we don't really need but where it's hard to get work done.

npmccallum replied:
I agree we need more time. I'd also like planning and retrospective to be back-to-back.

mbestavros replied:
Done in latest version. Resolving to simplify open feedback.

connorkuehl commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- What we accomplished in the last sprint (perhaps with a demo, if appropriate)
+- What didn’t get done, and why (blockers? More complex than anticipated? etc)
+
+If there is leftover work on our Active Sprint board, we should decide whether we should continue working on it or demote it to a backlog board. Then, we move onto the next sprint and do it all over again.
```
We should discuss why it didn't get completed. This might happen naturally as whoever is working on it explains blockers throughout the week, but broadly speaking understanding the consequences of our planning will help us plan better moving forward. There's a good chance we underestimated the difficulty, or perhaps there was a lot of scaffolding work that needed to fall into place before this one did that we didn't foresee. In both cases, we should regroom the issue and further decompose it if necessary.

MikeCamel replied:
"We" is dangerous here.  Again - there needs to be a discussion of roles here.  Until we have the roles described, we're going to struggle with what happens in each type of meeting.

blazebissar replied:
Agree with the above comments: without clear roles and responsibilities it is easy to repeat the story of [Everybody, Anybody, Somebody, and Nobody](https://www.lollydaskal.com/leadership/story-everybody-somebody-anybody-nobody/) or just talk in circles without reaching an output. I've seen that meetings tend to be ineffective without understood inputs, agendas/roles, and outputs. In this case, the outputs should also include how to improve the process itself (as mentioned below), how to address the issues experienced during the sprint, and the personnel responsible for each output task.

npmccallum replied:
Agreed. I want roles clearly defined.

lkatalin commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+### Active Sprint Board
+This board is where the team will track progress on current sprint tasks. **It should hold no backlog;** if an issue or PR is on this board, it is being actively worked on.
+
```
Moving issues and PRs between boards is easily done, right? I'm assuming so but I've never tried it myself. Ease of doing this could inform whether we want the component-specific backlog boards.

axelsimon replied:
At least for issues, it's just matter of picking which board to assign it to, as long as the issue and board you want are part of the same repo. Cross-repo transfer is more involved (but perfectly doable).

lkatalin commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Ready for review: where pull requests land by default (automatable with stock Github); indicates it’s ready to be reviewed by someone. Pull requests should move back here from Changes Requested if a fix has been applied and it’s ready for a re-review (not in stock Github, but automatable with Actions).
+- Changes Requested: PRs move here when a review is submitted that requests changes (automatable with stock Github).
+- Approved: PRs that have been approved go here, pending merge (automatable with stock Github).
+- Done: PRs that have been merged, and their associated issues, go here (automatable with stock Github).
```
Does "done" on the sprint board automatically move issues to "done" on the backlog boards? Are the issues duplicates of each other? I'm not 100% on how backlog and sprint stay in sync since we don't have pointers on GitHub boards....

lkatalin commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+A few things to consider:
+- Advantage: having project board be separate from backlog means we can promote issues to be part of the current sprint whenever we want to, and they can skip the backlog entirely if that’s desired.
+- Automation opportunity: If pull requests have an associated issue, don’t put the PR on the board -- instead, move the associated issue to the board (if not on it already) and treat the issue like the pull request, subject to the same movement rules as specified above. This would prevent us from having to look at the same thing twice, while still preserving a direct link to the PR in the context ...
```
> If pull requests have an associated issue

Would they ever not?

lkatalin replied:
I'm curious if this might address my question about duplicate issues between sprint / component boards. So the issue goes on the sprint board and the PR goes on the component board?

lkatalin commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+From there, automation will aid the process -- any newly assigned issues will:
+- be moved to the Active column in the backlog board
+- automatically added to the Assigned column of the Active Sprint board
```
Oh, so it does sound like there _are_ duplicate issues?

lkatalin commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+Next, we’ll have a look at all untriaged issues -- any that do not have a label, and are not part of a project. This will allow the whole team to discuss them and ensure they get moved to the appropriate backlog board and labeled correctly. This shouldn’t take more than 15 minutes, though that’s a soft limit -- if a good conversation is happening, it shouldn’t be cut off.
+
+Finally, if there's time, we'll dedicate some time to a quick review of our Current Sprint board. This will serve as an opportunity for developers to elaborate on any blockers and ask for reviews or re-reviews as needed. We should strive to do this every day if possible, but if prior discussion makes that infeasible, we don't _have_ to do it (there's always tomorrow). Project discussion is more i ...
```
Maybe it's just me, but it seems like a daily discussion of blockers / opportunity to ask questions and report progress is the biggest value of our current stand-up. Do we actively want this to change? If so, that could be fine (we still have the chat for that sort of thing) but I just wanted to point out that this description seems very different from what we're doing currently.

MikeCamel replied:
Agree - I'd expect individual reports to happen before untriaged issues discussion.

axelsimon replied:
Being able to flag any blockers and generate the opportunity to address them at some time after the meeting is one of the most valuable aspects of the daily stand-up, i think. So I agree with @lkatalin and @MikeCamel above.
To be specific, i think this aspect is more important than "what i have done, what i'm doing next" (although they are also valuable).

mbestavros replied:
What I was trying to express here was that those discussions around blockers (which are valuable, absolutely) should happen in the context of the issues themselves in the current sprint as opposed to just a quick status update.

If the current sprint board has everything we're working on anyway, we'd be duplicating status by talking about it in an individual report and then later on when we're looking at the board. That's redundant and a potential waste of time, IMO.

Of course, it's not hard at all to add personal reports back in. Just wanted to share some of my rationale.

lkatalin commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+- Implementing some of the automation exactly as specified may end up being a challenge. There's further discussion of this below.
+
+- Bisecting our backlog into different projects has its benefits, but it also precludes a nice, unified view of what needs to be done on the project. This would be addressable with a `backlog` label if we considered that a large-enough problem.
```
Yeah, I think using labels for more of the organization and dropping the different boards could make sense - currently a lot of our labels are redundant with the "projects." I don't have a strong opinion on this, though.

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+## Motivation
+
+The Enarx project needs to organize its work better, and in a more sustainable way than before. We need to do a better job planning for medium-term work and goals, and in a way that scales more effectively than our prior process.
```
```suggestion
The Enarx project needs to organize its work better, and in a more sustainable way than before. The project needs to do a better job planning for medium-term work and goals, and in a way that scales more effectively than our prior process. As we anticipate the project's separation into sub-teams working on specific aspects of Enarx to grow, we need to organise work in such a way that anyone can easily know what each team is working on right now, what are its priorities, and which team member is assigned to which task.
```

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+By default, we shouldn't do anything when an issue is first submitted. The lack of metadata will indicate it needs to be triaged. (If desired, we could post a quick welcome comment if they're not part of the core team, but that's optional.)
+
+Every standup, we'll look at all issues that are not part of a project and not labeled. This ensures that new issues always get a human response within 24 working hours, and that all issues are organized properly into the backlog -- or, if needed, promoted to the current sprint. It also allows the team to discuss all new issues as they come up.
```
I'm not sure we should do this, it sounds quite time consuming: actually qualifying an issue and ensuring everyone shares a common understanding of it takes time. Most issues don't need an answer within 24 hours. If they truly are urgent, they could be labelled "urgent". If they are security-related, there's another process for that. Lastly, any remaining issue *can* still be addressed shortly, if nothing else prevents it.

The usual process in agile sprints is to have a backlog grooming session which is used to get a set of stories ready to a point where they can be simply picked up for the next sprint.
This doesn't mean we can't discuss new issues as they come up, but usually the full conversation and especially the alignment of understanding and expectations would be kept for the backlog grooming session.

Lastly, "promoted to the current sprint" is something i would advise against. The point of a sprint is that it protects you against outside interference : the team decides what it will work on during the sprint and gets to delivering just that. Sprints are short, so anything else stakeholders want can actually wait until the next one, 95% of the time. You want to avoid setting a precedent of allowing issues opened halfway through the sprint to be accepted in the sprint, disrupting it.
There can still be exceptions, of course (as suggested in paragraph one) but in any case, they should be done on a one-in-one-out principle, as the team has already planned the sprint to do just as much as it can, no more, no less.

npmccallum replied:
@axelsimon My concern here is that new contributors who file issues or PRs get prompt engagement from the team. Let's say new contributor Bob files a PR immediately after the grooming meeting. Should he wait 2 weeks for his PR to get merged? Agile wasn't designed with open source in mind. So I'm looking for a tweak to agile retain open source distinctives.

MikeCamel replied:
An alternative, of course, is to have shorter sprints.  I'm not sure the benefits are worth it, however, but wanted to raise the option.

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+To start, I propose the following backlog projects:
+
+- Documentation
```
Ideally, documentation would not be a separate thing but built into each story/issue (and an agree-upon condition for considering something "done"), but that can also be considered to be a agile purist position :)

MikeCamel replied:
No, it's not purist: we need documentation to be part of each story/issue.  But there are also non-component documentation tasks which may need to be considered separately.

axelsimon replied:
Good point.

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Active: for issues that are actively being worked on as part of the current sprint.
+- Done: when things are done, they go here (automatable with stock Github). Mostly to ensure the “Active” column stays organized.
+
+In addition to inherently categorizing our backlog, this approach allows new "categories" to be moved around or created as needed -- and, if we decide to use organization-level projects for our sprints, allows the backlog to be bisected across repositories if wanted. This will support modularization as the project grows.
```
```
This will support modularization as the project grows.
```
:+1:

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- WASM
+
+A few things to consider:
+- Downside: issues that don’t fit squarely into a category don’t have a place to go. Maybe a "Miscellaneous" project?
```
This is where just treating each sub-board as a filtered view of the main, global board helps: any issue not specific to a sub-project is a general, Enarx issue.

MikeCamel commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+## Unresolved questions
+
+The most significant unresolved question does not involve process, but rather the automation we hope to pair with it: whether the Active Sprint board will live inside the main Enarx repository, or as an all-encompassing Enarx organization board.
```
Disagree: for me, the most signficant unresolved question revolves around role definitions and the responsibility of each role in the various steps in the process.

m-dombrowski replied:
Automation was my first flag early on and I see it addressed later in the text. I appreciate the breakdown of concepts in the document. Please forgive my ignorance, I must ask how much this alternate process will streamline efforts. More to the point, does this process reduce or increase overall administrative overhead in addition to engineer related tasks? If it does streamline, is there an associated risk of "paralysis by analysis?"

The document indicates a notable level of uncertainty regarding the automation of this process and a less centralized use of a platform. Is there an understood exit strategy if the transition fails (eg automation) and the project needs to revert to existing project management methods, in whole or in part?

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Assigned: this indicates a task that is currently being worked on as part of the sprint. Newly-added issues land here by default (automatable with stock Github).
+- Ready for review: where pull requests land by default (automatable with stock Github); indicates it’s ready to be reviewed by someone. Pull requests should move back here from Changes Requested if a fix has been applied and it’s ready for a re-review (not in stock Github, but automatable with Actions).
+- Changes Requested: PRs move here when a review is submitted that requests changes (automatable with stock Github).
+- Approved: PRs that have been approved go here, pending merge (automatable with stock Github).
```
I'm not sure what the Approved state gives us, i might be missing something.
Is there anything more to this state than describing PRs where everything is good, but simply haven't been merged yet? Would there be a reason to hold back and wait on merging said PRs if everything is good with them?

If not, it sounds like we can go directly to done, by simply merging any PR in "ready to review" which doesn't require any further changes.

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+This is meant to be a time to consider what work the team wants to accomplish in the next two-week sprint and assign relevant tasks to people. It should happen on Mondays and take about an hour, following the sprint retrospective (detailed later).
+
+It should start with a brief discussion about the team’s goals. This should be an open, but high-level, discussion that will help frame the kinds of issues we should prioritize. It shouldn’t take long -- no more than 15-20 minutes.
+
```
Very much agreed, aligning on what we want to achieve as a team over the coming sprint is both helpful for individual work, as you know why you are doing what you are doing, but also creates a better sense of having a common goal and puts said individual work in context.

MikeCamel replied:
As recently noted, what "the team" means here is becoming interesting.

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Automation opportunity: If pull requests have an associated issue, don’t put the PR on the board -- instead, move the associated issue to the board (if not on it already) and treat the issue like the pull request, subject to the same movement rules as specified above. This would prevent us from having to look at the same thing twice, while still preserving a direct link to the PR in the context ...
+
+### Sprint Planning
+With our backlog and sprint boards established, we can flesh out our first meeting: sprint planning.
```
In my experience of agile methodologies, the first meeting, which is a prerequisite to sprint planning, is backlog grooming.
The idea is:
1. backlog grooming: the team looks through the backlog of unrefined issues, and selects the next ones to groom. Any new issue that someone brings to the table as important is a good candidate, as are issues that are just the logical next step. The person who opened the issue explains it to the team, and a conversation regarding its importance, difficulty and, importantly, the "definition of done" occurs. The team then votes to assign "story points", which are arbitrary but serve as a way to compare team members' understanding and reveals any gap in comprehension. When a common value is reached, it's recorded on the story/issue and the issue is now one that can be picked for a sprint, during sprint planning.
- sprint planning: having properly groomed the stories/issues in the previous meeting, the team can simply pick a bunch of stories/issues for the sprint according to two things:
  - the goal of the sprint
  - the total number of points the stories/issues add up to
Sprint planning is also where, theoretically, stories get divided into smaller, immediately actionable tasks. In practise, thinking ahead to these tasks helps realise the complexity of the issue and therefore its point value (this should also help answer some of @lkatalin's questions!).

So normally, the sprint planning is a lot easier and faster, than the grooming.

So i suppose my question is: when is grooming of issues supposed to be done? I think your suggestion is during triage, so a little bit every day. I have no experience of doing that, and as mentioned in my comment re: triage, i have doubts regarding how it would work out given how time consuming it is.

lkatalin replied:
Thanks @axelsimon :) these comments are useful to read.

MikeCamel replied:
@axelsimon - I want to see roles and responsibilities noted.  Sometimes voting isn't appropriate, and the project might need guidance to hit strategic goals, for instance, which might prioritise particular sprints.  One might consider, for instance, a spring dedicated to docs or RFCs which might never make a popular vote, but which are vital to the project.  Product owners have a role here.

axelsimon commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+- Automation opportunity: If pull requests have an associated issue, don’t put the PR on the board -- instead, move the associated issue to the board (if not on it already) and treat the issue like the pull request, subject to the same movement rules as specified above. This would prevent us from having to look at the same thing twice, while still preserving a direct link to the PR in the context ...
+
+### Sprint Planning
+With our backlog and sprint boards established, we can flesh out our first meeting: sprint planning.
```
In my experience of agile methodologies, the first meeting, which is a prerequisite to sprint planning, is backlog grooming.
The idea is:
1. backlog grooming: the team looks through the backlog of unrefined issues, and selects the next ones to groom. Any new issue that someone brings to the table as important is a good candidate, as are issues that are just the logical next step. The person who opened the issue explains it to the team, and a conversation regarding its importance, difficulty and, importantly, the "definition of done" occurs. The team then votes to assign "story points", which are arbitrary but serve as a way to compare team members' understanding and reveals any gap in comprehension. When a common value is reached, it's recorded on the story/issue and the issue is now one that can be picked for a sprint, during sprint planning.
2. sprint planning: having properly groomed the stories/issues in the previous meeting, the team can simply pick a bunch of stories/issues for the sprint according to two things:
  - the goal of the sprint
  - the total number of points the stories/issues add up to
Sprint planning is also where, theoretically, stories get divided into smaller, immediately actionable tasks. In practise, thinking ahead to these tasks helps realise the complexity of the issue and therefore its point value (this should also help answer some of @lkatalin's questions!).

So normally, the sprint planning is a lot easier and faster, than the grooming.

So i suppose my question is: when is grooming of issues supposed to be done? I think your suggestion is during triage, so a little bit every day. I have no experience of doing that, and as mentioned in my comment re: triage, i have doubts regarding how it would work out given how time consuming it is.

axelsimon replied:
In my experience of agile methodologies, the first meeting, which is a prerequisite to sprint planning, is backlog grooming.
The idea is:
1. backlog grooming: the team looks through the backlog of unrefined issues, and selects the next ones to groom. Any new issue that someone brings to the table as important is a good candidate, as are issues that are just the logical next step. The person who opened the issue explains it to the team, and a conversation regarding its importance, difficulty and, importantly, the "definition of done" occurs. The team then votes to assign "story points", which are arbitrary but serve as a way to compare team members' understanding and reveals any gap in comprehension. When a common value is reached, it's recorded on the story/issue and the issue is now one that can be picked for a sprint, during sprint planning.
2. sprint planning: having properly groomed the stories/issues in the previous meeting, the team can simply pick a bunch of stories/issues for the sprint according to two things:
  - the goal of the sprint
  - the total number of points the stories/issues add up to
Sprint planning is also where, theoretically, stories get divided into smaller, immediately actionable tasks. In practise, thinking ahead to these tasks helps realise the complexity of the issue and therefore its point value (this should also help answer some of @lkatalin's questions!).

So normally, the sprint planning is a lot easier and faster, than the grooming.

So i suppose my question is: when is grooming of issues supposed to be done? I think your suggestion is during triage, so a little bit every day. I have no experience of doing that, and as mentioned in my comment re: triage, i have doubts regarding how it would work out given how time consuming it is.

blazebissar commented on concepts/00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,158 @@
+
+## Drawbacks
+
+- The team will need to attend a few more meetings. The daily standups, as currently designed, lose their focus on individual areas of the project per day; this was useful in case an external contributor was interested in one specific area of the project. However, we hope the added free-form discussion time will allow for relevant questions to be asked and talked about anyway.
```
As a newcomer, I actually think the free-form discussion is more valuable than the area-focus approach. The responsiveness and availability of the core team and the good use of chat make it easy to engage about a specific area as well.

Additionally, newcomer and external contributers are not present at most meetings and are not the majority of participants, so I think it is reasonable to structure meetings to maximally enable the core team and make adjustments as needed for new members when they do show up.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+ - **Planning**, a meeting representing the start of a new sprint
+ - **Daily Standup**, the team's daily meeting where, among other things,
+   **Triage** is performed.
+ - **Retrospective**, or "Retro" for short, a meeting representing the
```
Suggest change to "Review" - as discussed in the call, the retro is only part of the review.

mbestavros replied:
Sure, makes sense. I've updated to reflect the name change in my next push.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+### Triage
+
+Triage is where the team will categorize newly-filed issues and pull requests,
+and where stakeholders will be able to express needed urgency over those items.
```
Stakeholders are not defined.  This is important, as they are the ones making decisions about what to accept.  If we mean "Product Owner", then let's say so.  If others are involved, we should specify.

mbestavros replied:
Fixed in latest push.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+process for determining such issues is further detailed in the Retrospective
+meeting (detailed more below).
+
+With that done, the project's stakeholders now have the responsibility to
```
As noted above, the "stakeholders" are not defined.

mbestavros replied:
Fixed in latest push.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+follows:
+
+- 10-15 min: triage
+- 15-20 min: Board review/project discussion (however long is left after triage)
```
Where does discussion of "blockers" come in?  This is an important part of stand-ups.

mbestavros replied:
During board review, as detailed a paragraph or two back. I've put in a small reminder on that line in my next push to make it clear, though.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+
+Here's a proposed schedule:
+
+- 20 min: review issues/PRs from last sprint
```
Demos first, if for no other reason that we will sometimes invite visitors, and they shouldn't have to sit through the rest of the meeting.

mbestavros replied:
Agreed, good point. Fixed in my next push.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+  formatting fix, others might be a radical design change. Presently (and for
+  the foreseeable future), the Enarx project uses them exclusively to organize
+  tasks yet to be done, discussion topics, and more (even when not strictly
+  needed).
```
To be accepted, an issue **must** be well-scoped, and have a definition of "done".  Those writing issues are encouraged to specify or suggests tests to help the assignees to complete the task as expected.

Each issue should also be small enough to be accomplished within one Sprint: if not, it should be decomposed.  In the rare situations where this is not possible, sign-off by the Product Owner is required to spread it over multiple Sprints.

mbestavros replied:
I've added a section under "Planning" called "Acceptance Criteria" for this in my latest push.

MikeCamel commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+- every reviewer that has submitted a review has updated their review to
+  "Approved"
+
+The latter exists to ensure that all feedback from all reviewers is addressed
```
Do we want to require tests for code-based PRs?

mbestavros replied:
I suppose I assumed that was implied. Code tests are currently required to pass anyway. We can make that an explicit requirement for merge if desired. (Makes more sense thinking about it, especially in the context of potentially auto-merging some PRs.)

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+## Unresolved questions
+
+The most significant unresolved question does not involve process, but rather
+the automation we hope to pair with it: whether the project board will
+live inside the main Enarx repository, or as an all-encompassing Enarx
+organization board.
```
Isn't this question resolved now? I see under "Project Board" it claims that the board will be organization-level.

mbestavros replied:
You are right. I've updated it locally and will push it up soon.

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+Once a PR is in the `reviewed` state, automation will take care of updating
+assignees to reflect whose responsibility the PR is at the present time:
+
+- The reviewer that submitted the review is unassigned.
+- The PR author is assigned to the PR to address the reviewer's feedback.
```
Maybe this is implied, but it seems that the PR author should only be re-assigned to the PR if the reviewer did in fact request changes. Another option is the reviewer approved, in which case the re-assignment of the PR author is unnecessary. I think it would be annoying for a PR author to see they are re-assigned to an issue, only to check it and see they have no changes to make.

mbestavros replied:
Yeah, it might be helpful to update the wording here. The "reviewed" state really indicates "changes requested" I think.

I'll also note there is a separate state for explicit approval, and that doesn't involve any reassignment. It's true, though, that this doesn't handle a "review that doesn't request changes". In either case, the submitter doesn't need to do anything.

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+
+![PR workflow diagram](./pr-workflow.svg)
+
+As with every issue, every PR will be reviewed through a Triage, during which
+reviewers will be requested and assigned. At least two reviewers must be
```
Does this mean that if a PR comes in and you feel motivated to review it, you should *not* review it until the next day's triage? Or does it mean that you can review it and become and "optional reviewer", and other reviewers may be assigned the next day? The latter makes more sense to me.

mbestavros replied:
Yes, the latter.

The implication for someone submitting a "changes requested" review is that they're on board to keep reviewing the PR until their feedback is resolved. That shouldn't need to wait for a formal triage process. That's mostly there to ensure that PRs _have_ the necessary reviewers.

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+The sprint planning process is illustrated in the following flowchart:
+
+![Sprint Planning Diagram](./sprint-planning.svg)
+
```
This is a detail, but what do the diamonds vs. squares mean?

mbestavros replied:
They're standard (ish) ISO flowchart shapes: https://en.wikipedia.org/wiki/Flowchart

I've added an appendix that explains what they mean in the context of this RFC. Pushing soon.

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+The meeting will start with a triage, as described above. This exists to ensure
+no issues are uncategorized before the planning step. (This is necessary since
+sprint planning will supersede the day's standup and, consequently, its triage,
+as described later.)
+
+With the backlog fully up to date after the triage, the entire team will take
+a small amount of time to **nominate** issues from the area they work on that
+they think are most compelling. It should be each individual team member's
+responsibility to nominate issues of interest _before_ planning begins;
```
This is a bit buried, but I think is very important so that we don't get bogged down in going through every single backlog issue during nominations or have to wait while people make decisions. An inefficient process here would cut into our time for targeted discussion or make it likely that the meeting runs over-time.

In fact, I would argue that during triage, we should have the choice of moving non-urgent issues either to backlog or, if likely to be a priority for next sprint, then straight to "nominated"; this allows us to speed up our individual pre-planning nomination process by being able to quickly throw a small amount of issues from "nominated" into backlog if we want to de-prioritize them, rather than combing through a large amount of backlog issues to remember what we wanted to add to "nominated." 

If the above idea is too unorthodox for sprints, at the very least I think the responsibility described here that the development team moves issues into "nominated" before planning should be a semi-hard requirement. Last minute-nominations can make sense, but they should be the exception and the time limit on this should be enforced.

This way we have adequate time to discuss all the nominations and we minimize (individually and as a team) the process of searching through dozens of backlog issues to find a specific issue we know we want to nominate (but may have forgotten the exact wording of the title, etc.) while preparing for sprints.

mbestavros replied:
I personally like that idea. Maybe it can be a new state in the triage process flowchart if the issue is decided to be "not urgent".

lkatalin replied:
I didn't see where / if this was addressed, but maybe since it's a small detail it makes more sense to discuss as a team rather than put it in the document. Especially as I don't know others' thoughts on it.

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+that means the PR is still in review and likely just needs a bit more review
+time. Thus, it should stay as-is in the Under Review column.
+
+Beyond that, the rest of the meeting should be dedicated to demos or other
+high-level discussion about work accomplished.
```
Should demos be assigned as part of the sprint? Does the sprint take into account the time spent on creating a demo in addition to resolving issues? Do we care if we end up with zero demos or a very large number of demos for a sprint? Do we want to have bounds on this?

mbestavros replied:
Great question.

For the moment, I'd imagine we'd have to lump the cycles needed to produce a demo in with the overall task, or define making the demo as a separate task. In any case, this is something the team would have to discuss during the prioritization/bandwidth discussion part of sprint planning.

axelsimon replied:
> Should demos be assigned as part of the sprint? Does the sprint take into account the time spent on creating a demo in addition to resolving issues?

Normally, they should. Demos should really be just that, a demonstration of what has been accomplished during the sprint, so the whole team and stakeholders can come around what they have built and see progress. This is based on the idea that each sprint should deliver incremental value, and that every sprint as such always produces working code. We may be still a bit away from that, so i would loosen that requirement to just demonstrating what has been worked on and progress there, not adamantly requiring that it is a "usable increment".

lkatalin commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,542 @@
+a maximum of 24 working hours -- a desirable property for open source.
+
+Following triage, the scrum master will pull up the sprint board for an informal
+board review. This does not need to go over every single issue; rather, it
+should be an opportunity for team members to point out highlights or interesting
+updates to their tasks, and request help on blockers or otherwise.
```
I think it would be helpful to have a "show of hands" for this part of the meeting re: how many people want to talk about a highlight or blocker. The "round the horn" process we use currently has the advantage of making it immediately obvious how many people will want to say something (i.e., the total number in the meeting) and so pacing of the meeting flows from this naturally. If an unknown number of people have comments to make during this portion, we run into the danger of either 1) the meeting being cut short before everyone gets their say, or 2) the meeting running over-time. Therefore I propose a quick 5-second check before we move into this portion of how many people will bring up a discussion item, like a mini-agenda.

mbestavros replied:
Not a bad idea! I've added that in my latest revision.

mbestavros commented:
My latest push addressed @lkatalin's review.

> One additional thought I had (not inline) is: are the sprint boards archived in some way so that we can easily look back on what we prioritized and accomplished during previous sprints? I think this could be very useful both to assess our own work and show what we have done to others.

Given we use the same sprint board for every sprint, it'd be hard to preserve what _exactly_ was accomplished in a given sprint including states, etc. If we _really_ wanted to have that, we could maybe look at having a milestone for every sprint, but I'm not sure that's ideal.

For now, the best answer to what we've done is "what's been closed/merged".

mbestavros commented:
My latest push addresses @MikeCamel's feedback, specifically:

- refactored "Retrospective" to be "Review" (where the Retrospective process is a component);
- replaced "stakeholders" with "product owners" to be explicitly clear who those people are;
- added subheaders to Planning, and added a section on what an issue needs to be accepted.

I've also updated several of the diagrams with a better key and updated naming in some cases.

npmccallum commented:
@axelsimon @blazebissar @m-dombrowski @lkatalin @connorkuehl @MikeCamel 

I believe we have addressed all your comments and are nearly ready to merge. Can you please mark your comments as resolved where appropriate so that we can identify which issues still need to be addressed? Thank you!

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+
+![Sprint Retro diagram](./sprint-retro.svg)
+
+Any issues that don't have an associated PR are assumed to be "not started".
```
This assumes that issues are not code-related (outreach, documentation, mentoring) are never started. That being said, this can be manually sorted out, but there is some bias towards code-related issues here.

mbestavros replied:
The natural fix here is "Any issues that don't have an associated PR **and are still open**." Though you're still right that this doesn't communicate that "work has been done, but it's not quite shippable yet" -- though, then again, that's also a problem with issues that might be close to a PR, but not there yet.

The key thing here is a deliverable. For code issues, that's a PR. For non-code issues, that could be any number of other things. I can specify to make that a little more clear.

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+
+### Review
+
+The final meeting is **Review**, which represents the end of a
+sprint. It is a time for the whole team to reflect on work accomplished, demo
+new things, and discuss how to approach tasks that didn't get completed.
+
+Most of these topics will take the form of discussion. The only significant
+process to detail here is **Retrospective**, the workflow for reviewing issues
+and PRs from the last sprint.
```
I think calling this meeting **Review** is a good idea, however i take issue with calling **Retrospective** something that is disconnected from the [accepted meaning](http://retrospectivewiki.org/index.php?title=Agile_Retrospective_Resource_Wiki) of an agile retrospective, especially if we are telling folks we are basing our process on agile processes.

In light of this,
1. we should actually make time for engaging in retrospectives, as they are a good tool to ensure the "agile scaffolding" so to speak is working out for the team members and
2. call what is currently being termed "retrospective" an issue review, a card review, an end of sprint triage, or something like that.

mbestavros replied:
While perhaps not explicitly stated, I intended for point 1) to be addressed in the last 20 minutes of the meeting ("demos or other discussion" -- which should be simplified to "other discussion".)

I will clarify that those last 20 minutes should be for "retrospective".

As for 2)... I'm not really a fan of conflating "triage" with an "end of sprint triage". What about something like "end of sprint issue cleanup"?

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+Once an issue is filed, it will be categorized and discussed in Triage. If
+urgent, it's injected straight into the current sprint; if not, it gets pushed
+to the backlog. From there, the rest of the cycle will carry it through to
+completion: Planning will sooner or later land it as part of a sprint, where it
+will become a team member's action item to complete. At the end of the sprint,
```
> Planning will sooner or later land it as part of a sprint

Probably, but there is no guarantee of this: what happens if an issue keeps staying in the backlog?

Not needed right now, but we should probably eventually add a time to cull old issues that are not getting put in sprints, as they have become stale. If they are in fact not stale, why they are not getting added to sprints should in itself be a question worth answering, as it's a sign of some other problem.

mbestavros replied:
Good point. I will add something to the Future Possibilities section.

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+that means the PR is still in review and likely just needs a bit more review
+time. Thus, it should stay as-is in the Under Review column.
+
+Retrospective should happen immediately following demos (if there are any).
+Putting demos first allows those interested to see them without having to sit
+through the minutiae of project discussion.
```
This is good and makes sense, but makes even more sense if you consider/make explicit that all stakeholders should be invited to the demo: some of the stakeholders have no interest in the minutiae of project discussion but are very interested in seeing the progress made. We start with the demo to let them leave after.

Suggestion, add somewhere above the following:

> All stakeholders are invited to the demo, so they can get a sense of the progress made in the last sprint, or since they were last able to attend a demo.

mbestavros replied:
Added something to this effect.

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+
+A pull request should only be merged when:
+
+- all required code tests pass
+- it has two or more approving reviews
+- every reviewer that has submitted a review has updated their review to
+  "Approved"
```
All of these are good. However, while i realise these checks are more on the "tickbox" level and this may not fit here, there is no mention of documentation, neither here nor, to the best my reading, in the overall document.

We could possibly add something as simple as:

```suggestion
- all required code tests pass
- it has two or more approving reviews
- every reviewer that has submitted a review has updated their review to
  "Approved"
- any changes that require documentation to be understandable and/or maintainable have the required associated documentation
```

mbestavros replied:
Good point. I've addressed this in my latest version.

npmccallum replied:
```suggestion
- all required code tests pass
- it has two or more reviews in the `APPROVED` state
- it has no reviews in the `CHANGES_REQUESTED` state
- it has no reviews in the `DISMISSED` state
```

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+#### Acceptance criteria
+
+It is the product owner's responsibility to only accept issues that meet
+acceptance criteria::
```
```suggestion
acceptance criteria:
```

mbestavros replied:
Fixed, thanks!

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+It is the product owner's responsibility to only accept issues that meet
+acceptance criteria::
+
+- it must be well-scoped, and have a clear definition of completion
```
Cf. comment on line 420 and following.

```suggestion
- it must be well-scoped, and have a clear definition of completion, including documentation if needed
```

mbestavros replied:
Added in my latest revision.

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+- it must be well-scoped, and have a clear definition of completion
+- it must be doable in a single sprint
+
+If an issue cannot be accomplished in a single sprint, it should be decomposed.
+In the rare scenario that an issue cannot be decomposed, it needs to be signed
+off by the product owners to be included in a sprint.
+
+Additionally, the product owners must be careful to accept a reasonable amount
+of work from Proposed and not overload the team members. The purpose of this
+step is not to express that all work is important; that's counterproductive and
+doesn't accomplish anything.
+
```
I really like both of these.

axelsimon commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,593 @@
+- With automation, this process has potential to minimize the busywork that
+  project developers have to deal with, and only require human judgment when it
+  is needed.
+- Running through a triage every day ensures issues are promptly reviewed by the
+  team after they are submitted.
```
I have not seen been done before, but i like the idea. I'm looking forward to seeing how this works out, as i think the idea of spreading out a backlog grooming, doing a bit every day, is quite smart.

mbestavros commented:
I've addressed most of @axelsimon's review, though I haven't yet refactored "Retrospective" to be something new.

npmccallum commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,612 @@
+- accepted: issues the team should work on next sprint (planning)
+- assigned: issues actively being worked on by an assigned team member
+- under review: all submitted PRs go here while awaiting review
+- done: all issues and PRs go here when done or merged, respectively
```
We need one additional board: `triage`. The bot should put all open issues that are not in the project board into the `triage` column.

ueno commented on 00001-enarx-sprint-framework/README.md:
```diff
@@ -0,0 +1,628 @@
+
+### Tracking Work
+
+As a Github-native project, Enarx categorizes its work with the two major
```
Sorry for nitpicking, but I would suggest spelling it as "GitHub".

mbestavros replied:
Fixed!
